helix-core/src/syntax/span.rs

```diff
@@ -67,6 +71,79 @@ pub fn span_iter(spans: Vec<Span>) -> impl Iterator<Item = HighlightEvent> {
     }
 }
 
+impl SpanIter {
+    fn start_span(&mut self, span: Span) {
+        use HighlightEvent::*;
+
+        debug_assert!(span.start <= span.end);
+        self.event_queue
+            .push_back(HighlightStart(Highlight(span.scope)));
+        self.range_ends.push(span.end);
+    }
+
+    fn process_range_end(&mut self, end: usize) -> HighlightEvent {
+        use HighlightEvent::*;
+
+        let start = replace(&mut self.cursor, end);
+        if start != end {
+            debug_assert!(start < end);
+            self.event_queue.push_back(HighlightEnd);
+            Source { start, end }
+        } else {
+            HighlightEnd
+        }
+    }
+
+    // Any subslices that end before intersect span needs to be subsliced, consume the
+    // left part of the subslice and leave the right.
+    fn partition_spans_at(&mut self, intersect: usize) {
+        let first_partioned_span = self.spans[self.index];
+
+        let mut i = self.index;
+        while let Some(span) = self.spans.get_mut(i) {
+            if span.start != self.cursor || span.end < intersect {
+                break;
+            }
+
+            let mut partinoed_span = *span;
+            partinoed_span.end = intersect;
+            span.start = intersect;
+
+            self.start_span(partinoed_span);
+            i += 1;
+        }
+
+        let subslices = i - self.index;
+
+        // When spans are subsliced, the span Vec may need to be re-sorted
+        // because the `range.start` may now be greater than some `range.start`
+        // later in the Vec. This is not a classic "sort": we take several
+        // shortcuts to improve the runtime so that the sort may be done in
+        // time linear to the cardinality of the span Vec. Practically speaking
+        // the runtime is even better since we only scan from `self.index` to
+        // the first element of the Vec with a `range.start` after this range.
+        let mut after = None;
+        let intersect_pos = Span {
+            start: intersect,
+            ..first_partioned_span
+        };
+        while let Some(span) = self.spans.get(i) {
+            if span <= &intersect_pos {
+                after = Some(i);
+                i += 1;
+            } else {
+                break;
+            }
+        }
```

Comment:
I opted for the `loop { match { Some(range) if .. { .. }, _ => break, } }` pattern since I thought the `while let` version was less clear that we're only looking for exactly one pattern. What do you think, is the `while let` version more idiomatic?

Reply:
I personally like the while loop better, because it looks less cluttered.
I personally always prefer `if` to `match` if there is just one case that I actually care about.
Its also closer to my ideal solution which would be a let_chain whenever that becomes available on a rust version we can use: `while let Some(span) = self.spans.get(i) && span <= &intersect_pos `

The `while` loop is just a standard vec iterator so when scanning the code its immediately) obvious what it does and I can focus on the content of the loop (where anything interesting is happening).

However thinking about this more this can actually be solved with an iterator which makes it much clearer.
This also inspired me to rename some variables to make the behaviour of `partition_spans_at` much clearer.
These changes are in 30563c4dd5c5d727c41af39abf665ebee2065ba4. Let me know what you think